spirv_msl.cpp

```diff
@@ -18403,6 +18868,189 @@ void CompilerMSL::emit_block_hints(const SPIRBlock &)
 {
 }
 
+void CompilerMSL::emit_mesh_outputs()
+{
+	auto &mode = get_entry_point();
+
+	// predefined thread count or zero, if specialization constant is in use
+	uint32_t num_invocaions = 0;
+	if (mode.workgroup_size.id_x == 0 && mode.workgroup_size.id_y == 0 && mode.workgroup_size.id_z == 0)
+	{
+		num_invocaions = mode.workgroup_size.x * mode.workgroup_size.y * mode.workgroup_size.z;
+	}
+
+	{
+		statement("threadgroup_barrier(mem_flags::mem_threadgroup);");
+		statement("if (spv_primitive_count == 0)");
+		begin_scope();
+		statement("return;");
+		end_scope();
+		statement("spvMesh.set_primitive_count(spv_primitive_count);");
+	}
+
+	if (mesh_out_per_vertex != 0)
+	{
+		auto &type_vert = get<SPIRType>(mesh_out_per_vertex);
+		if (num_invocaions < mode.output_vertices)
+		{
+			statement("for (uint spvI = gl_LocalInvocationIndex, spvThreadCount = "
+			          "(gl_WorkGroupSize.x*gl_WorkGroupSize.y*gl_WorkGroupSize.z); spvI < ",
+			          mode.output_vertices, "; spvI += spvThreadCount)");
+		}
+		else if (num_invocaions != mode.output_vertices)
+		{
+			statement("if (gl_LocalInvocationIndex < ", mode.output_vertices, ")");
+		}
+		begin_scope();
+		if (num_invocaions >= mode.output_vertices)
+		{
+			statement("const uint spvI = gl_LocalInvocationIndex;");
+		}
+		statement("spvPerVertex spvV = {};");
+		for (uint32_t index = 0; index < uint32_t(type_vert.member_types.size()); ++index)
+		{
+			uint32_t orig_var =
```

Comment:
This path is a bit too naive for more complex objects. E.g.

```
#version 450
#extension GL_EXT_mesh_shader : require
layout(max_vertices = 3, max_primitives = 1, triangles) out;
layout(local_size_x = 1) in;

out gl_MeshPerVertexEXT
{
	invariant vec4 gl_Position;
} gl_MeshVerticesEXT[3];

layout(location = 0) out float foos[3][4];

layout(location = 4) out Foo
{
	float bar;
} bars[3];

void main()
{
	SetMeshOutputsEXT(3, 1);
	gl_MeshVerticesEXT[0].gl_Position = vec4(1.0);
	gl_MeshVerticesEXT[1].gl_Position = vec4(1.0);
	gl_MeshVerticesEXT[2].gl_Position = vec4(1.0);
	gl_PrimitiveTriangleIndicesEXT[0] = uvec3(0, 1, 2);
	foos[0][0] = 10.0;
	foos[1][1] = 20.0;
	foos[2][2] = 20.0;
	bars[0].bar = 4.0;
	bars[1].bar = 5.0;
	bars[2].bar = 6.0;
}
```

```
    for (uint spvI = gl_LocalInvocationIndex, spvThreadCount = (gl_WorkGroupSize.x*gl_WorkGroupSize.y*gl_WorkGroupSize.z); spvI < 3; spvI += spvThreadCount)
    {
        spvPerVertex spvV = {};
        spvV.gl_Position = gl_MeshVerticesEXT[spvI].gl_Position;
        spvV.foos_0 = foos[spvI];
        spvV.foos_1 = foos[spvI];
        spvV.foos_2 = foos[spvI];
        spvV.foos_3 = foos[spvI];
        spvV.bars_bar = bars[spvI].bar;
        spvMesh.set_vertex(spvI, spvV);
    }
```

it doesn't seem to lower arrayed objects.

I feel like there should probably be a way to reuse the existing lambda stuff to lower the output writes. That might be something I have to look into once the rest of the implementation is in an acceptable state.